packages/contracts/contracts/Unirep.sol

```diff
@@ -378,47 +398,61 @@ contract Unirep is IUnirep, zkSNARKHelper, Hasher, VerifySignature {
 
     /**
      * A user spend reputation via an attester, the non-zero nullifiers will be processed as a negative attestation
-     * @param input The epoch key and its proof and the public signals
+     * publicSignals[0: maxReputationBudget ] = [ reputationNullifiers ]
+     * publicSignals[maxReputationBudget    ] = [ epoch ]
+     * publicSignals[maxReputationBudget + 1] = [ epochKey ]
+     * publicSignals[maxReputationBudget + 2] = [ globalStateTree ]
+     * publicSignals[maxReputationBudget + 3] = [ attesterId ]
+     * publicSignals[maxReputationBudget + 4] = [ proveReputationAmount ]
+     * publicSignals[maxReputationBudget + 5] = [ minRep ]
+     * publicSignals[maxReputationBudget + 6] = [ proveGraffiti ]
+     * publicSignals[maxReputationBudget + 7] = [ graffitiPreImage ]
+     * @param publicSignals The public signals of the reputation proof
+     * @param proof The The proof of the reputation proof
      */
-    function spendReputation(ReputationProof memory input) external payable {
-        bytes32 proofNullifier = Hasher.hashReputationProof(input);
+    function spendReputation(
+        uint256[] memory publicSignals,
```

Comment:
Because this is external So I think we can use datacall at here and another methods. I not sure I remember it right so Please check it! Thanks!

Reply:
I found that event cannot be emitted correctly if using `calldata`. (some data will miss)
like the error here https://app.circleci.com/pipelines/github/Unirep/Unirep/28/workflows/ed3273a3-cf74-4f64-8154-33f6ca5ea85f/jobs/172?invite=true#step-105-396
I think we need more time to see how to deal with this